File: leetcode/102. Binary Tree Level Order Traversal/main.ts
// Solved by @kitanoyoru
// https://leetcode.com/problems/binary-tree-level-order-traversal/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

const levelOrder = (root: TreeNode | null): number[][] => {
  let hm: number[][] = []
  if (!root) {
    return hm
  }

  let q: TreeNode[] = []
  q.push(root)

  while (q.length) {
    let count = q.length
    let level: number[] = []

    while (count--) {
      let node = q.shift()!
      level.push(node.val)
      if (node.left) {
        q.push(node.left)
      }
      if (node.right) {
        q.push(node.right)
      }
    }

    hm.push(level)
  }

  return hm
}

export {}

File: leetcode/102. Binary Tree Level Order Traversal/main.rs
// Solved by @kitanoyoru
// https://leetcode.com/problems/binary-tree-level-order-traversal/submissions/

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut ans = vec![];

        let mut q = vec![];
        if let Some(root) = root {
            q.push(root);
        }

        while !q.is_empty() {
            let mut level = vec![];
            let mut nq = vec![];

            for node in q.iter() {
                level.push(node.borrow().val);
                if let Some(left) = node.borrow_mut().left.take() {
                    nq.push(left);
                }
                if let Some(right) = node.borrow_mut().right.take() {
                    nq.push(right);
                }
            }

            ans.push(level);
            q = nq;
        }

        ans
    }
}
